browser/src/control/Control.ContextToolbar.ts

```diff
@@ -0,0 +1,296 @@
+/* -*- js-indent-level: 8 -*- */
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * L.control.ContextToolbar - definition of context toolbar
+ * This toolbar should appear over selection when related selection is done using mouse
+ */
+
+class ContextToolbar {
+	container: HTMLElement;
+	builder: any;
+	map: any;
+	initialized: boolean = false;
+	// roughly twice the height(76px) of default context toolbar in each direction from boundary
+	disappearingBoundary: number = 150; // px
+
+	constructor(map: any) {
+		this.map = map;
+		this.builder = new L.control.notebookbarBuilder({
+			windowId: -2,
+			mobileWizard: this,
+			map: this.map,
+			cssClass: 'notebookbar',
+			useSetTabs: true,
+		});
+		this.container = L.DomUtil.createWithId(
+			'div',
+			'context-toolbar',
+			document.body,
+		);
+		L.DomUtil.addClass(this.container, 'notebookbar horizontal');
+	}
+
+	showContextToolbar(pos: Point): void {
+		if (!this.initialized) {
+			this.builder.build(this.container, this.getWriterTextContext(), false);
+			this.initialized = true;
+		}
+
+		this.builder.map.on('jsdialogaction', this.onJSAction, this);
+		this.builder.map.on('jsdialogupdate', this.onJSUpdate, this);
+		document.addEventListener('pointermove', this.pointerMove);
+		this.changeOpacity(1);
+		this.showHideToolbarImp(true, pos);
+	}
+
+	calculateOpacity(e: PointerEvent): number {
+		const clientRect: DOMRect = this.container.getBoundingClientRect();
```

Comment:
costly operation, can we avoid it somehow?
best to run in layouting task inside showHideToolbarImp what will save us from unwanted frame render out of the schedule

Reply:
yes indeed its costly i have tried to avoid multiplication and division especially, but any new equation is welcomed. meantime i moved it to layout task.